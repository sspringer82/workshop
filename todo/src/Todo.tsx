import React from 'react';
import { Todo as TodoType } from './Todo.interface';

const Todo: React.FC = () => {
  const todo: TodoType = {
    id: '1',
    title: 'Get up',
    done: true,
  };

  return (
    <div>
      <div>Id: {todo.id}</div>
      <div>Title: {todo.title}</div>
      <div>Done: {todo.done}</div>
    </div>
  );
};

export default Todo;
