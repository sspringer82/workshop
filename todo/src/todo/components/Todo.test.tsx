import React from 'react';
import { render, fireEvent, wait } from '@testing-library/react';
import Todo from './Todo';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';

describe('<Todo>', () => {
  describe('add', () => {
    let add: (a: number, b: number) => number;

    beforeEach(() => {
      // arrange
      add = (a: number, b: number): number => a + b;
    });

    it('should add 1 and 1 and return 2', () => {
      // act
      const result = add(1, 1);
      // assert
      expect(result).toBe(2);
      expect(result).not.toBe(3);
      expect(result).toBeGreaterThan(1);
      expect(result).toBeLessThan(3);
    });
    it('should add 2 and 2 and return 4', () => {
      // act
      const result = add(2, 2);
      // assert
      expect(result).toBe(4);
    });
  });

  it('Render the component', () => {
    const todo = { id: 'aaa', title: 'bbb', done: false };
    const { getByTestId } = render(
      <Router>
        <Todo todo={todo} onStatusChange={() => {}} />
      </Router>
    );

    const title = getByTestId('title').textContent;
    expect(title).toBe('bbb');
  });

  it('should change the status', async () => {
    const cb = jest.fn();

    const todo = { id: 'aaa', title: 'bbb', done: false };
    const { getByTestId } = render(
      <Router>
        <Todo todo={todo} onStatusChange={cb} />
      </Router>
    );
    const changeStatusButton = getByTestId('changeStatus');
    await wait(() => {
      fireEvent.click(changeStatusButton);
    });

    expect(cb).toHaveBeenCalled();
    expect(cb).toHaveBeenCalledWith({ id: 'aaa', title: 'bbb', done: false });
  });

  it('renders correctly', () => {
    const todo = { id: 'aaa', title: 'bbb', done: false };
    const tree = renderer
      .create(
        <Router>
          <Todo todo={todo} onStatusChange={() => {}} />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
