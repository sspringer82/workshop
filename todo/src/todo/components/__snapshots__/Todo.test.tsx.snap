// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Todo> renders correctly 1`] = `
<div
  className="MuiPaper-root MuiCard-root sc-AykKD ewcSlM MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="sc-AykKE cBTOBV"
  >
    <div
      className="MuiCardContent-root sc-AykKC gGkBwA"
    >
      <h2
        className="MuiTypography-root MuiTypography-h5"
      >
        <a
          href="/form/aaa"
          onClick={[Function]}
        >
          <span
            data-testid="title"
          >
            bbb
          </span>
          (
          aaa
          )
        </a>
      </h2>
      <div
        data-testid="changeStatus"
        onClick={[Function]}
      >
        ‚ùå
         
      </div>
      <button
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 15.31L23.31 12 20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69zM12 18V6c3.31 0 6 2.69 6 6s-2.69 6-6 6z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
